crucible-wasm/src/Lang/Crucible/Wasm.hs

```diff
@@ -0,0 +1,316 @@
+, execScript
+, emptyScriptState
+, WasmExt
+, extImpl
+, wasmIntrinsicTypes
+) where
+
+import Control.Lens hiding (Empty, (:>),Index )
+import Control.Monad
+import Control.Monad.Trans
+
+import qualified Data.Map as Map
+import qualified Data.Sequence as Seq
+import qualified Data.Text.Lazy as TL
+import qualified Data.BitVector.Sized as BV
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Parameterized.Map as MapF
+import           Data.Word
+
+import Data.Parameterized.Context as Ctx
+import Data.Parameterized.Some
+
+import qualified Language.Wasm as Wasm
+import qualified Language.Wasm.Structure as Wasm
+
+import Lang.Crucible.Backend
+import Lang.Crucible.FunctionHandle
+import Lang.Crucible.Simulator
+import Lang.Crucible.Types
+
+import qualified Lang.Crucible.LLVM.MemModel.Generic as G
+
+import Lang.Crucible.Wasm.Extension
+import Lang.Crucible.Wasm.Instantiate
+import Lang.Crucible.Wasm.Memory
+import Lang.Crucible.Wasm.Translate
+
+import What4.Interface
+
+wasmIntrinsicTypes :: IsSymInterface sym => IntrinsicTypes sym
+wasmIntrinsicTypes =
+  MapF.insert (knownSymbol @"Wasm_Mem") IntrinsicMuxFn $
+  MapF.empty
+
+
+data ScriptState =
+  ScriptState
+  { scriptStore :: Store
+  , scriptNamespace :: Namespace
+  , scriptLoadedModules :: Namespace
+  , scriptCurrentModule :: Maybe (Maybe Wasm.Ident, Wasm.ValidModule, ModuleInstance)
+  }
+
+emptyScriptState :: ScriptState
+emptyScriptState =
+  ScriptState
+  { scriptStore = emptyStore
+  , scriptNamespace = emptyNamespace
+  , scriptLoadedModules = emptyNamespace
+  , scriptCurrentModule = Nothing
+  }
+
+type WasmOverride p sym = OverrideSim p sym WasmExt (RegEntry sym UnitType)
+
+execScript :: [Wasm.Command] -> ScriptState -> WasmOverride p sym EmptyCtx UnitType ScriptState
+execScript script ss = foldM (flip execCommand) ss script
+
+execCommand :: Wasm.Command -> ScriptState -> WasmOverride p sym EmptyCtx UnitType ScriptState
+execCommand (Wasm.ModuleDef mdef) ss =
+  do halloc <- use (stateContext . to simHandleAllocator)
+     let st = scriptStore ss
+     let ns = scriptNamespace ss
+     ((nm,vm,im), st') <- runInstM st (defineModule halloc ns mdef)
+
+     initGlobals im st'
+     initMemories im st'
+
+     mapM_ registerFnBinding =<< liftIO (translateFunctions im st')
+     liftIO (print (instDataSegs im))
+     mapM_ writeDataSegment (instDataSegs im)
+
+     executeStart im
+
+     let lmods' = case nm of
+                    Nothing  -> scriptLoadedModules ss
+                    Just nm' -> namespaceInsertModule nm' im (scriptLoadedModules ss)
+
+     return ss{ scriptStore = st'
+              , scriptLoadedModules = lmods'
+              , scriptCurrentModule = Just (nm,vm,im)
+              }
+
+execCommand (Wasm.Assertion a) ss =
+  do execAssertion a ss
+     return ss
+
+execCommand (Wasm.Action a) ss =
+  do _ <- execAction a ss
+     return ss
+
+execCommand c ss =
+  do liftIO $ putStrLn $ unwords ["Command not implemented", show c]
```

Comment:
That's a good question.  I'm coming around to the idea that we should perhaps make a distinction between `panic` and `unimplemented`.  In my mind, `panic` means the programmer has thought about this and believes that various explicit or implicit invariants mean that the situation cannot occur, and it is a programming mistake bug if the situation occurs.  Whereas, `unimplemented` or `unsupported` is explicitly an acknowledgement that valid inputs can drive the program into this situation, but we just haven't written the necessary code to deal with it.

Operationally, I think `unimplemented` should probably raise an appropriate (potentially non-fatal) exception instead of printing the panic message and immediately exiting.

Reply:
I agree that there's definitely a semantic differentiation, but is there a runtime distinction (i.e. can it be a non-fatal/recoverable exception if it's truly unimplemented)?  If so, then yes, let's make an `unimplemented`, otherwise we can just make sure the panic message clarifies the distinction (or I guess we can make an `unimplemented` that does the latter).

Reply:
Well... for something like SAW, I guess I'd rather see some useful "unimplemented" message and end up back at the SAW prompt rather than baling out of the executable altogether like `panic` does.

Reply:
My overall thinking relative to all the panic notes below is leaving the right breadcrumbs for both the user and the developer to know who to tell and where to fix it when these unimplemented/todo/whatever things get encountered. 
 
As context: Comments in the code, `print` statements that might get lost or don't suggest there needs to be a report, or `error` calls that might simply result in alternatives or quietly unsatisfiable branches are mostly passive and could result in invalid analyses or a lot of debugging to figure out why something isn't working right, which is why I tend to default to a fairly agressive reaction like `panic`.